tests/estimators/online/joint/test_ecm_joint.py

```diff
@@ -0,0 +1,243 @@
+from typing import List
+
+import numpy as np
+import pytest
+
+from asoh.models.ecm.advancedSOH import ECMASOH
+from asoh.models.ecm.ins_outs import ECMInput, ECMMeasurement
+from asoh.models.ecm.transient import ECMTransientVector
+from asoh.models.ecm.simulator import ECMSimulator
+from asoh.estimators.online.joint.ecm.unscented import ECMJointUKF
+
+
+def cycle_protocol(rng, asoh: ECMASOH, start_time: float = 0.0) -> List[ECMInput]:
+    """
+    Function to return list of ECMInputs corresponding to a new cycle
+
+    Args:
+        asoh: A-SOH object to be used
+        start_time: time at which to start this cycle in seconds
+
+    Returns:
+        ecm_inputs: list of ECMInput objects corresponding to this cycle
+    """
+    # initialize main lists to be used
+    timestamps = []
+    currents = []
+
+    # Get the Qt and CE, which will help determine the currents
+    qt = asoh.q_t.amp_hour
+    ce = asoh.ce
+
+    # Now, let's choose rest durations and C-rates
+    rest0, rest1, rest2 = rng.integers(low=5, high=10, size=3)
+    rest0 *= 60
+    rest1 *= 2 * 60
+    rest2 *= 60
+    dischg_rate, chg_rate = np.array([1.5, 0.5]) + rng.random(size=2)
+
+    def update_timestamps_curr(duration: float, curr: float) -> None:
+        """
+        Helper to populate timestamps and currents.
+
+        Args:
+            duration: duration of segment in seconds
+            curr: value of current to be used.
+        """
+        # specify binds to nearest variables
+        nonlocal currents
```

Comment:
Nice! I didn't know about non-local as an alternative to global.

Reply:
I did not either... praise the stackoverflow overlords!